File: scoring/scoring.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Dict, List, Tuple

from core.models import ItemRecord


@dataclass
class ScoreResult:
    field_scores: Dict[str, float]
    overall: float


def _score_date(item_date: datetime, config: Dict[str, int], urgency: int, urgency_override: int) -> float:
    if urgency == urgency_override:
        return 5.0
    recent_days = config.get("recent_days", 7)
    mid_days = config.get("mid_days", 30)
    days_old = (datetime.now() - item_date).days
    if days_old <= recent_days:
        return 1.0
    if days_old <= mid_days:
        return 3.0
    return 5.0


def _score_cost(cost: float, bands: List[Dict[str, float]]) -> float:
    for band in bands:
        max_val = band.get("max")
        if max_val is None:
            return float(band["score"])
        if cost <= float(max_val):
            return float(band["score"])
    return 1.0


def _weighted_average(pairs: List[Tuple[float, float]]) -> float:
    numerator = sum(score * weight for score, weight in pairs)
    denominator = sum(weight for _, weight in pairs) or 1.0
    return numerator / denominator


def score_item(item: ItemRecord, weights_config: Dict) -> ScoreResult:
    weights = weights_config.get("weights", {})
    date_cfg = weights_config.get("date_scoring", {})
    cost_bands = weights_config.get("cost_bands", [])
    urgency_override = weights_config.get("urgency_override", 5)

    scores = {
        "date": _score_date(item.date, date_cfg, item.urgency, urgency_override),
        "cost": _score_cost(item.cost, cost_bands),
        "urgency": float(item.urgency),
        "value": float(item.value),
        "price_comp": float(item.price_comp),
        "effect": float(item.effect),
    }

    pairs = [(scores[key], float(weights.get(key, 1.0))) for key in scores]
    overall = _weighted_average(pairs)
    scores["overall"] = overall
    return ScoreResult(field_scores=scores, overall=overall)

File: src/scoring.rs
use crate::config::WeightsConfig;
use crate::models::ItemRecord;
use chrono::{Datelike, Local};

pub struct ScoreResult {
    pub field_scores: FieldScores,
    pub overall: f64,
}

pub struct FieldScores {
    pub date: f64,
    pub cost: f64,
    pub urgency: f64,
    pub value: f64,
    pub price_comp: f64,
    pub effect: f64,
}

pub fn score_item(item: &ItemRecord, cfg: &WeightsConfig) -> ScoreResult {
    let date_score = score_date(item, cfg);
    let cost_score = score_cost(item, cfg);
    let scores = FieldScores {
        date: date_score,
        cost: cost_score,
        urgency: item.urgency as f64,
        value: item.value as f64,
        price_comp: item.price_comp as f64,
        effect: item.effect as f64,
    };

    let pairs = vec![
        (scores.date, cfg.weights.date),
        (scores.cost, cfg.weights.cost),
        (scores.urgency, cfg.weights.urgency),
        (scores.value, cfg.weights.value),
        (scores.price_comp, cfg.weights.price_comp),
        (scores.effect, cfg.weights.effect),
    ];
    let overall = weighted_avg(&pairs);
    ScoreResult {
        field_scores: scores,
        overall,
    }
}

fn score_date(item: &ItemRecord, cfg: &WeightsConfig) -> f64 {
    if item.urgency == cfg.urgency_override {
        return 5.0;
    }
    let days_old = (Local::now().date_naive() - item.date.date_naive()).num_days();
    if days_old <= cfg.date_scoring.recent_days {
        1.0
    } else if days_old <= cfg.date_scoring.mid_days {
        3.0
    } else {
        5.0
    }
}

fn score_cost(item: &ItemRecord, cfg: &WeightsConfig) -> f64 {
    for band in &cfg.cost_bands {
        if let Some(max) = band.max {
            if item.cost <= max {
                return band.score;
            }
        } else {
            return band.score;
        }
    }
    1.0
}

fn weighted_avg(pairs: &[(f64, f64)]) -> f64 {
    let num: f64 = pairs.iter().map(|(s, w)| s * w).sum();
    let den: f64 = pairs.iter().map(|(_, w)| w).sum();
    if den == 0.0 { 0.0 } else { num / den }
}
